File: google/chromeos/moblab/v1beta1/google-cloud-chromeos-moblab-v1beta1-java/gapic-google-cloud-chromeos-moblab-v1beta1-java/src/test/java/com/google/chromeos/moblab/v1beta1/BuildServiceClientTest.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.chromeos.moblab.v1beta1;

import static com.google.chromeos.moblab.v1beta1.BuildServiceClient.ListBuildsPagedResponse;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GaxGrpcProperties;
import com.google.api.gax.grpc.testing.LocalChannelProvider;
import com.google.api.gax.grpc.testing.MockGrpcService;
import com.google.api.gax.grpc.testing.MockServiceHelper;
import com.google.api.gax.rpc.ApiClientHeaderProvider;
import com.google.api.gax.rpc.InvalidArgumentException;
import com.google.api.gax.rpc.StatusCode;
import com.google.common.collect.Lists;
import com.google.longrunning.Operation;
import com.google.protobuf.AbstractMessage;
import com.google.protobuf.Any;
import io.grpc.StatusRuntimeException;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutionException;
import javax.annotation.Generated;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

@Generated("by gapic-generator-java")
public class BuildServiceClientTest {
  private static MockServiceHelper mockServiceHelper;
  private static MockBuildService mockBuildService;
  private BuildServiceClient client;
  private LocalChannelProvider channelProvider;

  @BeforeClass
  public static void startStaticServer() {
    mockBuildService = new MockBuildService();
    mockServiceHelper =
        new MockServiceHelper(
            UUID.randomUUID().toString(), Arrays.<MockGrpcService>asList(mockBuildService));
    mockServiceHelper.start();
  }

  @AfterClass
  public static void stopServer() {
    mockServiceHelper.stop();
  }

  @Before
  public void setUp() throws IOException {
    mockServiceHelper.reset();
    channelProvider = mockServiceHelper.createChannelProvider();
    BuildServiceSettings settings =
        BuildServiceSettings.newBuilder()
            .setTransportChannelProvider(channelProvider)
            .setCredentialsProvider(NoCredentialsProvider.create())
            .build();
    client = BuildServiceClient.create(settings);
  }

  @After
  public void tearDown() throws Exception {
    client.close();
  }

  @Test
  public void listBuildsTest() throws Exception {
    Build responsesElement = Build.newBuilder().build();
    ListBuildsResponse expectedResponse =
        ListBuildsResponse.newBuilder()
            .setNextPageToken("")
            .addAllBuilds(Arrays.asList(responsesElement))
            .build();
    mockBuildService.addResponse(expectedResponse);

    ModelName parent = ModelName.of("[BUILD_TARGET]", "[MODEL]");

    ListBuildsPagedResponse pagedListResponse = client.listBuilds(parent);

    List<Build> resources = Lists.newArrayList(pagedListResponse.iterateAll());

    Assert.assertEquals(1, resources.size());
    Assert.assertEquals(expectedResponse.getBuildsList().get(0), resources.get(0));

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    ListBuildsRequest actualRequest = ((ListBuildsRequest) actualRequests.get(0));

    Assert.assertEquals(parent.toString(), actualRequest.getParent());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void listBuildsExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      ModelName parent = ModelName.of("[BUILD_TARGET]", "[MODEL]");
      client.listBuilds(parent);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void listBuildsTest2() throws Exception {
    Build responsesElement = Build.newBuilder().build();
    ListBuildsResponse expectedResponse =
        ListBuildsResponse.newBuilder()
            .setNextPageToken("")
            .addAllBuilds(Arrays.asList(responsesElement))
            .build();
    mockBuildService.addResponse(expectedResponse);

    String parent = "parent-995424086";

    ListBuildsPagedResponse pagedListResponse = client.listBuilds(parent);

    List<Build> resources = Lists.newArrayList(pagedListResponse.iterateAll());

    Assert.assertEquals(1, resources.size());
    Assert.assertEquals(expectedResponse.getBuildsList().get(0), resources.get(0));

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    ListBuildsRequest actualRequest = ((ListBuildsRequest) actualRequests.get(0));

    Assert.assertEquals(parent, actualRequest.getParent());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void listBuildsExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      String parent = "parent-995424086";
      client.listBuilds(parent);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void checkBuildStageStatusTest() throws Exception {
    CheckBuildStageStatusResponse expectedResponse =
        CheckBuildStageStatusResponse.newBuilder()
            .setIsBuildStaged(true)
            .setStagedBuildArtifact(BuildArtifact.newBuilder().build())
            .setSourceBuildArtifact(BuildArtifact.newBuilder().build())
            .build();
    mockBuildService.addResponse(expectedResponse);

    BuildArtifactName name =
        BuildArtifactName.of("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]");

    CheckBuildStageStatusResponse actualResponse = client.checkBuildStageStatus(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    CheckBuildStageStatusRequest actualRequest =
        ((CheckBuildStageStatusRequest) actualRequests.get(0));

    Assert.assertEquals(name.toString(), actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void checkBuildStageStatusExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      BuildArtifactName name =
          BuildArtifactName.of("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]");
      client.checkBuildStageStatus(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void checkBuildStageStatusTest2() throws Exception {
    CheckBuildStageStatusResponse expectedResponse =
        CheckBuildStageStatusResponse.newBuilder()
            .setIsBuildStaged(true)
            .setStagedBuildArtifact(BuildArtifact.newBuilder().build())
            .setSourceBuildArtifact(BuildArtifact.newBuilder().build())
            .build();
    mockBuildService.addResponse(expectedResponse);

    String name = "name3373707";

    CheckBuildStageStatusResponse actualResponse = client.checkBuildStageStatus(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    CheckBuildStageStatusRequest actualRequest =
        ((CheckBuildStageStatusRequest) actualRequests.get(0));

    Assert.assertEquals(name, actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void checkBuildStageStatusExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      String name = "name3373707";
      client.checkBuildStageStatus(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void stageBuildTest() throws Exception {
    StageBuildResponse expectedResponse =
        StageBuildResponse.newBuilder()
            .setStagedBuildArtifact(BuildArtifact.newBuilder().build())
            .build();
    Operation resultOperation =
        Operation.newBuilder()
            .setName("stageBuildTest")
            .setDone(true)
            .setResponse(Any.pack(expectedResponse))
            .build();
    mockBuildService.addResponse(resultOperation);

    BuildArtifactName name =
        BuildArtifactName.of("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]");

    StageBuildResponse actualResponse = client.stageBuildAsync(name).get();
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    StageBuildRequest actualRequest = ((StageBuildRequest) actualRequests.get(0));

    Assert.assertEquals(name.toString(), actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void stageBuildExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      BuildArtifactName name =
          BuildArtifactName.of("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]");
      client.stageBuildAsync(name).get();
      Assert.fail("No exception raised");
    } catch (ExecutionException e) {
      Assert.assertEquals(InvalidArgumentException.class, e.getCause().getClass());
      InvalidArgumentException apiException = ((InvalidArgumentException) e.getCause());
      Assert.assertEquals(StatusCode.Code.INVALID_ARGUMENT, apiException.getStatusCode().getCode());
    }
  }

  @Test
  public void stageBuildTest2() throws Exception {
    StageBuildResponse expectedResponse =
        StageBuildResponse.newBuilder()
            .setStagedBuildArtifact(BuildArtifact.newBuilder().build())
            .build();
    Operation resultOperation =
        Operation.newBuilder()
            .setName("stageBuildTest")
            .setDone(true)
            .setResponse(Any.pack(expectedResponse))
            .build();
    mockBuildService.addResponse(resultOperation);

    String name = "name3373707";

    StageBuildResponse actualResponse = client.stageBuildAsync(name).get();
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockBuildService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    StageBuildRequest actualRequest = ((StageBuildRequest) actualRequests.get(0));

    Assert.assertEquals(name, actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void stageBuildExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockBuildService.addException(exception);

    try {
      String name = "name3373707";
      client.stageBuildAsync(name).get();
      Assert.fail("No exception raised");
    } catch (ExecutionException e) {
      Assert.assertEquals(InvalidArgumentException.class, e.getCause().getClass());
      InvalidArgumentException apiException = ((InvalidArgumentException) e.getCause());
      Assert.assertEquals(StatusCode.Code.INVALID_ARGUMENT, apiException.getStatusCode().getCode());
    }
  }
}

File: google/chromeos/moblab/v1beta1/google-cloud-chromeos-moblab-v1beta1-ruby/test/google/cloud/chromeos/moblab/v1beta1/build_service_client_test.rb
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "minitest/autorun"
require "minitest/spec"

require "google/gax"

require "google/cloud/chromeos/moblab"
require "google/cloud/chromeos/moblab/v1beta1/build_service_client"
require "google/chromeos/moblab/v1beta1/build_service_services_pb"
require "google/longrunning/operations_pb"

class CustomTestError_v1beta1 < StandardError; end

# Mock for the GRPC::ClientStub class.
class MockGrpcClientStub_v1beta1

  # @param expected_symbol [Symbol] the symbol of the grpc method to be mocked.
  # @param mock_method [Proc] The method that is being mocked.
  def initialize(expected_symbol, mock_method)
    @expected_symbol = expected_symbol
    @mock_method = mock_method
  end

  # This overrides the Object#method method to return the mocked method when the mocked method
  # is being requested. For methods that aren't being tested, this method returns a proc that
  # will raise an error when called. This is to assure that only the mocked grpc method is being
  # called.
  #
  # @param symbol [Symbol] The symbol of the method being requested.
  # @return [Proc] The proc of the requested method. If the requested method is not being mocked
  #   the proc returned will raise when called.
  def method(symbol)
    return @mock_method if symbol == @expected_symbol

    # The requested method is not being tested, raise if it called.
    proc do
      raise "The method #{symbol} was unexpectedly called during the " \
        "test for #{@expected_symbol}."
    end
  end
end

class MockBuildServiceCredentials_v1beta1 < Google::Cloud::Chromeos::Moblab::V1beta1::Credentials
  def initialize(method_name)
    @method_name = method_name
  end

  def updater_proc
    proc do
      raise "The method `#{@method_name}` was trying to make a grpc request. This should not " \
          "happen since the grpc layer is being mocked."
    end
  end
end

describe Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient do

  describe 'list_builds' do
    custom_error = CustomTestError_v1beta1.new "Custom test error for Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient#list_builds."

    it 'invokes list_builds without error' do
      # Create request parameters
      formatted_parent = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.model_path("[BUILD_TARGET]", "[MODEL]")

      # Create expected grpc response
      next_page_token = ""
      total_size = 705419236
      builds_element = {}
      builds = [builds_element]
      expected_response = {
        next_page_token: next_page_token,
        total_size: total_size,
        builds: builds
      }
      expected_response = Google::Gax::to_proto(expected_response, Google::Chromeos::Moblab::V1beta1::ListBuildsResponse)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::ListBuildsRequest, request)
        assert_equal(formatted_parent, request.parent)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:list_builds, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("list_builds")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          response = client.list_builds(formatted_parent)

          # Verify the response
          assert(response.instance_of?(Google::Gax::PagedEnumerable))
          assert_equal(expected_response, response.page.response)
          assert_nil(response.next_page)
          assert_equal(expected_response.builds.to_a, response.to_a)
        end
      end
    end

    it 'invokes list_builds with error' do
      # Create request parameters
      formatted_parent = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.model_path("[BUILD_TARGET]", "[MODEL]")

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::ListBuildsRequest, request)
        assert_equal(formatted_parent, request.parent)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:list_builds, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("list_builds")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1beta1 do
            client.list_builds(formatted_parent)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'check_build_stage_status' do
    custom_error = CustomTestError_v1beta1.new "Custom test error for Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient#check_build_stage_status."

    it 'invokes check_build_stage_status without error' do
      # Create request parameters
      formatted_name = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.build_artifact_path("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]")

      # Create expected grpc response
      is_build_staged = true
      expected_response = { is_build_staged: is_build_staged }
      expected_response = Google::Gax::to_proto(expected_response, Google::Chromeos::Moblab::V1beta1::CheckBuildStageStatusResponse)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::CheckBuildStageStatusRequest, request)
        assert_equal(formatted_name, request.name)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:check_build_stage_status, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("check_build_stage_status")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          response = client.check_build_stage_status(formatted_name)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.check_build_stage_status(formatted_name) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes check_build_stage_status with error' do
      # Create request parameters
      formatted_name = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.build_artifact_path("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]")

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::CheckBuildStageStatusRequest, request)
        assert_equal(formatted_name, request.name)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:check_build_stage_status, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("check_build_stage_status")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1beta1 do
            client.check_build_stage_status(formatted_name)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'stage_build' do
    custom_error = CustomTestError_v1beta1.new "Custom test error for Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient#stage_build."

    it 'invokes stage_build without error' do
      # Create request parameters
      formatted_name = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.build_artifact_path("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]")

      # Create expected grpc response
      expected_response = {}
      expected_response = Google::Gax::to_proto(expected_response, Google::Chromeos::Moblab::V1beta1::StageBuildResponse)
      result = Google::Protobuf::Any.new
      result.pack(expected_response)
      operation = Google::Longrunning::Operation.new(
        name: 'operations/stage_build_test',
        done: true,
        response: result
      )

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::StageBuildRequest, request)
        assert_equal(formatted_name, request.name)
        OpenStruct.new(execute: operation)
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:stage_build, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("stage_build")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          response = client.stage_build(formatted_name)

          # Verify the response
          assert_equal(expected_response, response.response)
        end
      end
    end

    it 'invokes stage_build and returns an operation error.' do
      # Create request parameters
      formatted_name = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.build_artifact_path("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]")

      # Create expected grpc response
      operation_error = Google::Rpc::Status.new(
        message: 'Operation error for Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient#stage_build.'
      )
      operation = Google::Longrunning::Operation.new(
        name: 'operations/stage_build_test',
        done: true,
        error: operation_error
      )

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::StageBuildRequest, request)
        assert_equal(formatted_name, request.name)
        OpenStruct.new(execute: operation)
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:stage_build, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("stage_build")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          response = client.stage_build(formatted_name)

          # Verify the response
          assert(response.error?)
          assert_equal(operation_error, response.error)
        end
      end
    end

    it 'invokes stage_build with error' do
      # Create request parameters
      formatted_name = Google::Cloud::Chromeos::Moblab::V1beta1::BuildServiceClient.build_artifact_path("[BUILD_TARGET]", "[MODEL]", "[BUILD]", "[ARTIFACT]")

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Chromeos::Moblab::V1beta1::StageBuildRequest, request)
        assert_equal(formatted_name, request.name)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1beta1.new(:stage_build, mock_method)

      # Mock auth layer
      mock_credentials = MockBuildServiceCredentials_v1beta1.new("stage_build")

      Google::Chromeos::Moblab::V1beta1::BuildService::Stub.stub(:new, mock_stub) do
        Google::Cloud::Chromeos::Moblab::V1beta1::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::Chromeos::Moblab.new(version: :v1beta1)

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1beta1 do
            client.stage_build(formatted_name)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end
end